sjsonnet/src/sjsonnet/Val.scala

```diff
@@ -221,37 +244,65 @@ object Val{
       allKeys
     }
 
-    @inline def hasKeys = !getAllKeys.isEmpty
+    @inline def hasKeys: Boolean = {
+      val m = if (static || `super` != null) getAllKeys else value0
```

Comment:
Notably, I don't call `getValue0` here or in the other new branches added below: it turns out that `value0` is guaranteed to be defined for non-static objects, so `getValue0` is only effective for static objects.

When I started working on `sjsonnet`, I found the invariants on some of the `Obj` internals to be somewhat confusing. I'm considering adding a big block comment at the top of this class to document my current understanding. I think we especially need documentation for the invariants / constraints on static objects.